### ISSUE TITLE
Count blueprint for models

### ISSUE THREAD

**Issue by ghaiklor** [balderdashy/sails#2542]
Hi guys.

**The problem**
Need make route that will be mapped to each model and realize get count of records of model.

**Why?**
Sometimes for mobile devs it can be helpful for getting total counts of some data. Like *how many users already registered*, etc...

**Where?**
I think it can be done with /model/count route or smth like that.

**Solution**
Add `count.js` file to `actions` folder with code like below
```javascript
var actionUtil = require('../actionUtil');

module.exports = function getCount (req, res) {
  var Model = actionUtil.parseModel(req),
      criteria = actionUtil.parseCriteria(req);

  Model.count(criteria, function(error, response) {
    if (error) return res.serverError(error);
    res.ok({count: response});
  });
};
```

What you think about this ?

**Comment by ghaiklor** [comment]
BTW, I can make pull request for this if you think that this is the great idea.

**Comment by randallmeeker** [comment]
https://gist.github.com/RWOverdijk/f64d2c4e5bab4922e861#file-count-js

**Comment by ghaiklor** [comment]
@randallmeeker I mean, that will be great if this feature will be in Sails from the box. Why we need always write own blueprint if it can be integrated in Sails blueprints by default?

**Comment by randallmeeker** [comment]
@ghaiklor I can't imagine there is any particular reason, other than it has not been done yet. I would submit your pull request. All my blueprints are usually re written anyway. Blueprints out of the box don't really work. I also have pagination in my blue prints which return counts already as well.

https://github.com/randallmeeker/SailsBluePrintActions/tree/master/pagination

**Comment by ghaiklor** [comment]
@randallmeeker for me most of default blueprints is working as I expect and this fine. That's why I'm trying to ask guys implement this feature. But your variant great too. I should think about this.

**Comment by animedbz16** [comment]
This request is fairly similar to one that I had proposed and worked with @randallmeeker:
https://github.com/balderdashy/sails/issues/1730

Basically ensuring that every API request would return a content-range HTTP header with the requests.

This way if you performed a search or had paginated results, you could at least have the content range requested and a total count returned in the same request.

At the point when I requested this, we were really testing out possibly switching over to Sails for one application, but ended up going in another direction for that app since we are pretty far along in development with it so I didn't have time to submit a PR

We using sails for a new app, but haven't had much time to go back and address that to submit a PR.

I think having this functionality would be great to be included in the default blueprints.

**Comment by ghaiklor** [comment]
Ok, guys. I'll create PR when I have time and it will be done.
@animedbz16 so are you propose make blueprint that returns count of records in headers?

**Comment by hinok** [comment]
I decided not to create new issue because my problem is related to this.
I'm completly new sailor in the deep blue oceans and my ship needs total count of returned collection in http headers. I saw solution of @randallmeeker and changed it a little.

```javascript
/**
 * Module dependencies
 */
var util = require('util'),
  actionUtil = require('../../node_modules/sails/lib/hooks/blueprints/actionUtil');


/**
 * Find Records
 *
 *  get   /:modelIdentity
 *   *    /:modelIdentity/find
 *
 * An API call to find and return model instances from the data adapter
 * using the specified criteria.  If an id was specified, just the instance
 * with that unique id will be returned.
 *
 * Optional:
 * @param {Object} where       - the find criteria (passed directly to the ORM)
 * @param {Integer} limit      - the maximum number of records to send back (useful for pagination)
 * @param {Integer} skip       - the number of records to skip (useful for pagination)
 * @param {String} sort        - the order of returned records, e.g. `name ASC` or `age DESC`
 * @param {String} callback - default jsonp callback param (i.e. the name of the js function returned)
 */

module.exports = function findRecords(req, res) {

  // Look up the model
  var Model = actionUtil.parseModel(req);

  // If an `id` param was specified, use the findOne blueprint action
  // to grab the particular instance with its primary key === the value
  // of the `id` param.   (mainly here for compatibility for 0.9, where
  // there was no separate `findOne` action)
  if (actionUtil.parsePk(req)) {
    return require('./findOne')(req, res);
  }

  // Lookup for records that match the specified criteria
  var query = Model.find()
    .where(actionUtil.parseCriteria(req))
    .limit(actionUtil.parseLimit(req))
    .skip(actionUtil.parseSkip(req))
    .sort(actionUtil.parseSort(req));
  // TODO: .populateEach(req.options);
  query = actionUtil.populateEach(query, req);
  query.exec(function found(err, matchingRecords) {
    if (err) return res.serverError(err);

    // Only `.watch()` for new instances of the model if
    // `autoWatch` is enabled.
    if (req._sails.hooks.pubsub && req.isSocket) {
      Model.subscribe(req, matchingRecords);
      if (req.options.autoWatch) {
        Model.watch(req);
      }
      // Also subscribe to instances of all associated models
      _.each(matchingRecords, function(record) {
        actionUtil.subscribeDeep(req, record);
      });
    }

    Model.count(actionUtil.parseCriteria(req)).exec(function(error, total) {
        if (err) {
          return res.serverError(err);
        }

        res.set('Total-Count', total);
        res.ok(matchingRecords);
      });

  });
};
```

It almost works. Almost because now I always get count of collection even if my response is just an empty array.
I use sailsjs 0.10.5 + mysql

Example
---
I have endpoint `/event` and just one record in db.
I set parameter `skip=1` and make a call.
I get empty array (it's correct) but Model.count() still returns 1.

**Comment by ArtworkAD** [comment]
:+1: been waiting for the count blueprint to be integrated soooo long. Currently I am using this https://gist.github.com/RWOverdijk/f64d2c4e5bab4922e861#file-count-js

The downside is that we have to manually create routes for this blueprint as described here http://stackoverflow.com/questions/22273789/crud-blueprint-overriding-in-sailsjs

**Comment by ghaiklor** [comment]
@ArtworkAD I have made a little research around blueprints, so my guess that we need count only for `find` blueprint. Am I right ?
Also, why are you using custom overriding? Just create `api/blueprints/find.js` file with your custom blueprint. This fact, that you create `find.js` in `api/blueprints` means that you override default `find.js` blueprints. I'm using this one:

```javascript
var actionUtil = require('sails/lib/hooks/blueprints/actionUtil');

module.exports = function (req, res) {
    if (actionUtil.parsePk(req)) {
        return require('./findOne')(req, res);
    }

    var Model = actionUtil.parseModel(req),
        where = actionUtil.parseCriteria(req),
        limit = actionUtil.parseLimit(req),
        skip = actionUtil.parseSkip(req),
        sort = actionUtil.parseSort(req),
        query = Model.find().where(where).limit(limit).skip(skip).sort(sort);

    query = actionUtil.populateEach(query, req);
    query.exec(function (error, records) {
        if (error) {
            return res.serverError(error);
        }

        Model
            .count(where)
            .exec(function (error, count) {
                if (error) {
                    return res.serverError(error);
                }

                var metaInfo = {
                    start: skip,
                    end: skip + limit,
                    limit: limit,
                    total: count,
                    criteria: where
                };

                res.set('Content-Range', metaInfo.start + '-' + metaInfo.end + '/' + metaInfo.total);

                return res.ok(records, null, null, metaInfo);
            });
    });
};
```

Also you can check out my generator-sails-rest-api for yeoman - https://github.com/ghaiklor/generator-sails-rest-api. Stuff like `count` already implemented there and many other cool stuff. I had created this generator for myself, but now it's available open-source.

**Comment by ArtworkAD** [comment]
@ghaiklor thank you, just tested, it works great!

**Comment by CWyrtzen** [comment]
Did the PR resolve this issue? Can we close it up?

**Comment by ArtworkAD** [comment]
@CWyrtzen yes its resolved. You can close it up :smile:

**Comment by CWyrtzen** [comment]
gratzi!

**Comment by ghaiklor** [comment]
@CWyrtzen emm... nope. PR is not referenced to this issue.

**Comment by ArtworkAD** [comment]
Why not ?

**Comment by ghaiklor** [comment]
@ArtworkAD because this issue about blueprints and returning total count of records. You even didn't read issue description :+1:

**Comment by ArtworkAD** [comment]
@ghaiklor oh I did, I am not sure why you still want another solution when there is this PR https://github.com/balderdashy/sails/issues/2712 

You combine this with a custom find blueprint (see here https://gist.github.com/ArtworkAD/3fca7f386c971f73843d) and a response for a find request will have meta information in the header:

    start: skip,
    end: skip + limit,
    limit: limit,
    total: count,
    criteria: where

this is useful for pagination but you can also use it to get just the total count. I am comfortable with this solution.

**Comment by ghaiklor** [comment]
@ArtworkAD with this issue I want to make it by default in Sails. That's why this issue is still opened.

**Comment by CWyrtzen** [comment]
here ya go!

**Comment by mikedevita** [comment]
I am using [this](https://github.com/randallmeeker/SailsBluePrintActions/tree/master/pagination/inBody) blueprint override, I personally prefer the override and pushing into the `res` via `meta: {}` my question is how can I get this to work for associations? I need to "meta" information for any associations that have been populated for that model.

something like this would be awesome...

```javascript
[{
  info: {
    start: 0,
    end: 30,
    limit: 30,
    total: 3,
    criteria: {}
  },
  name: 'checklist1',
  items: {
    info: {
      start: 0,
      end: 30,
      limit: 30,
      total: 3,
      criteria: {}
    },
    statuses: [
      { ... },
      { ... },
      { ... },
      { ... },
    ]
  }
}];
```

**Comment by sailsbot** [comment]
Thanks for posting, @ghaiklor.  I'm a repo bot-- nice to meet you!

It has been 30 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!
